### PR TITLE
inductor: add native GQA to sliding window attention

### PR DESCRIPTION
**Pull request by ani300**
## Summary

- keep sliding-window K/V blocks at their native Hkv head count and use the broadcast-aware rank-5 matmul path added for SDPA
- preserve the explicit Hkv/group axes in SWA fallback tiling and cover batch > 1 mask broadcasting plus Gemma 3/4 production shapes
- apply the remaining dgrove-oss review suggestions from #4277 and #4423 with cost-model rationale, validation tests, and public API documentation

## Review follow-ups

From #4277:
- document that the low-KV-head 2 MiB target is motivated by Gemma 4 12B/26B global attention
- document the query-only native-GQA work-division rule in the SDPA selector

From #4423:
- parameterize every public SWA validation guard, including mask shape/dtype/device
- document the public schema, runtime-mask contract, compact-cache performance requirement, native GQA behavior, and compile-only restriction
- retain the detailed cache-position-first layout explanation already added to the test helper
- retain the SWA-specific measured tiling policy that answered the earlier 512-token tuning question

#4423 also contains a shared compiler fix outside SWA: `register_operation_after_graph_edit` prevents operation-name reuse after late graph edits, and its call sites in graph editing/LX context switching plus regression coverage are part of that predecessor PR.

## Testing

- `SENCORES=32 python -m pytest tests/inductor/test_sliding_window_attention.py tests/inductor/test_kv_window.py tests/inductor/test_swa_tiling.py -q` (155 passed)
- rebased-main smoke coverage: 19 focused unit tests and 4 compiled native-GQA/Gemma 4 device tests passed
- `pre-commit run --files ...` (all hooks passed)

### CODE REVIEW

**Review by dgrove-oss — APPROVED** [review]
Summary

Reworks sliding-window attention to keep K/V at their native Hkv head count throughout (cache, kv_window, and the online-softmax loop), using the pre-existing broadcast-aware spyre.batched_matmul (already used by full SDPA) instead of materializing Hq-sized expanded copies of K/V. Also folds in review follow-ups from #4277 and #4423: documents the low-KV-head 2 MiB tiling target, adds explicit validation guards with tests for every public SWA parameter, and adds docs/ coverage for the op.

Review

The core change is well-executed and internally consistent. I traced the shape flow through the whole path: kv_window's fake-tensor sizing now correctly uses key.shape[1] (native Hkv) instead of the passed num_heads; _windowed_attention unflattens query to [B, Hkv, group, Lq, D] and unsqueezes the mask/K/V window chunks at the matching axis (dim 2) so the rank-5 broadcast lines up correctly for both matmuls; the final flatten(1, 2) correctly restores [B, Hq, Lq, D] output shape only when use_gqa is set. The num_head_tiles=(1 if num_heads != num_kvheads else ...) tiling rule in _select_swa_tiling matches full SDPA's existing native-GQA rule, and the new test_swa_tiling.py cases (num_heads=12, num_kvheads=3) confirm this produces num_head_tiles=1 as expected. register_operation_after_graph_edit and batched_matmul referenced in the PR body are both confirmed pre-existing (from #4423), not part of this diff, matching how the PR describes them.

BLOCKERs

None found.

SUGGESTIONs

- torch_spyre/_inductor/decompositions.py:1372-1376: k_blk.unsqueeze(2)/v_blk.unsqueeze(2) are only applied if use_gqa; worth a one-line comment noting this axis must match query's/attention_mask's unsqueeze axis above (dim 2) — the correctness here rests on three separate call sites agreeing on the same axis number, which isn't obvious from any single site in isolation.

QUESTIONs

None — the design and validation are clear from the docstrings and new guard tests.

Test Coverage

Strong. New coverage spans all three affected layers:
- test_sliding_window_attention.py: a parametrized test_validation_guards covering all 10 new/existing guard conditions (including the new nondivisible_heads check), plus test_prefill_batch_gqa specifically targeting the batch>1 mask-broadcast case called out as load-bearing in the comment.
- test_kv_window.py: test_gqa_fake_preserves_native_kv_heads and the renamed test_gqa_key_window_preserves_native_kv_heads directly assert the op no longer materializes Hq-sized output — verified the reference helper _reference_window was correctly simplified (no longer calls _expand_kv) since the op itself no longer expands.
- test_swa_tiling.py: updated expectations for num_head_tiles=1 across all fallback-tiling test cases, plus a new test_unknown_gqa_geometry_preserves_native_head_axes.

CI/DCO

DCO: pass. Single commit, signed off correctly.
CI: 10 passing, 145 pending (freshly pushed), 1 skipping, 0 failing at time of review.
License headers: present and correct on all modified files (note: test_swa_tiling.py carries a 2026 copyright year, consistent with today's date).

Verdict

APPROVE
